### DISCUSSION TITLE
Toluidine Blue Image Analysis on AxonDeepSeg

### DISCUSSION THREAD

**Discussion by SPACEADITYA** [neuropoly/axondeepseg#614] [Help needed]
Hi All, 

I just downloaded AxonDeepSeg through the virtual python environment, and I've tested a couple of EM images with the program. I'm very happy to see that I get great segmentation with these images. However, when I try with other EM or Toluidine Blue stained images, I am unable to get a result. For example, I am able to successfully upload my images into fsleyes. However, when I click on "Apply ADS Prediction Model," and I am prompted for pixel size, I click ok and the program just doesn't segment my axons/myelin. 

Could this be an issue with the pixel size associated with my Image? If so, how do I determine the right pixel size for my image before inputting them into fsleyes? I am working with sciatic nerve cross sections stained with Toluidine blue. 

Finally, for the "select the modality," I don't have the traditional "TEM" or "SEM" options. I have attached a screenshot of what I am seeing below. I would appreciate any help with this!

<img width="1303" alt="image" src="https://user-images.githubusercontent.com/101671235/158465201-69edd128-4251-49a7-8c06-c68619c9c69e.png">

**Comment by hermancollin** [comment]
@SPACEADITYA thank you for reaching out. Regarding your modality question, what you see are the model names. 
- SEM: model_seg_rat_axon-myelin_sem 
- TEM: model_seg_mouse_axon-myelin_tem
- BF: model_seg_rat_axon-myelin_bf

As for the model not applying, maybe @Stoyan-I-A has seen a similar behavior in the past? What pixel size did you use?

**Comment by mathieuboudreau** [comment]
> and I am prompted for pixel size, I click ok and the program just doesn't segment my axons/myelin.

@SPACEADITYA do you enter a pixel size? Could you look in the terminal (where you launched FSLeyes) to see if there are any error messages there after you try segmenting the image?

**Reply by SPACEADITYA** [reply]
Hi All, 

Thank you so much for your responses. The modality makes much more sense now. If I am running Toluidine blue stained sections (not SEM or TEM), should I run that with BF (model_seg_rat_axon-myelin_bf)? My sections are from mice sciatic nerve. 

Also, I went ahead and checked in my terminal for any error messages when I was (1) uploading my tiff into the environment (using the GUI) and (2) when I clicked "Apply ADS Prediction Model."  I have attached screenshots of both error messages here. 

Error 1: 
![image](https://user-images.githubusercontent.com/101671235/158497927-4cb2a980-12d0-4df3-a998-66bfae878c2e.png)

Error 2: 
![image](https://user-images.githubusercontent.com/101671235/158497948-5b5cf63e-2812-41b7-bab8-2785336781c8.png)

Perhaps it's an image sizing issue since these are cropped images from a larger nerve section. So far I have tried analysing the image in ImageJ and making sure the pixel size there matches what I input into AxonDeepSeg when I am prompted (after clicking "Apply ADS Prediction Model" button). 

I'd appreciate any clarification! Thank You!

**Reply by mathieuboudreau** [reply]
@SPACEADITYA thanks for the additional detail!

I think the easiest way for us to replicate your issue and debug it is if you could provide us with a sample image (you can email it if you'd rather not post it here publicly) and resolution of an image that is giving you trouble (and if it exists, one pair that is working correctly).

As for your errors, the first error is not actually an error but a normal warning, and error 2 signals us that there may be a resampling issue, possible that your image has too few pixels (or resolution is too small/large). If this is the case, the fix should be easy but it would help for us to try with your image directly.

>The modality makes much more sense now. If I am running Toluidine blue stained sections (not SEM or TEM), should I run that with BF (model_seg_rat_axon-myelin_bf)? My sections are from mice sciatic nerve.

Typically, you should pick the model that was trained on images that has a contrast most similar to your own images. Although you say that your images are blue stained, the images you've uploaded are black and white and have a contrast closer to TEM (dark myelin, white axons), so I would try model_seg_mouse_axon-myelin_tem first, and then maybe the bf as a test comparison.

**Reply by SPACEADITYA** [reply]
Hi Mathieu, 

Thank you for your reply! I think it was a Pixel issue, and when I converted the image in ImageJ from 16bit to 8bit, and reuploaded it, the program worked. 

Overall my lab is interested in acquiring morphometrics on large tissue cross sections. The AxonDeepSeg program works very well on it already; however, when I try to optimise and click "Find Axons," the area it labels as Axon starts spilling out of the desired area. I have attached images below of what I am describing. 

Would you recommend I try training a different algorithm with ivadomed? Or continue using fsleyes and just try using the pencil and eraser tools to optimise. 

Initial Output: 
<img width="617" alt="Screenshot 2022-03-16 at 12 28 05 PM" src="https://user-images.githubusercontent.com/101671235/158639186-b2c37c3a-4fb0-4f34-8561-1835f8482306.png">

Zoomed Initial Output: 
<img width="885" alt="Screenshot 2022-03-16 at 12 28 42 PM" src="https://user-images.githubusercontent.com/101671235/158639312-b465ca0b-93bd-472a-aa45-cc42b971f34d.png">

When I try rerunning with myelin manually drawn in: 
<img width="855" alt="Screenshot 2022-03-16 at 12 29 44 PM" src="https://user-images.githubusercontent.com/101671235/158639551-348c64b0-d9fd-4fd3-a4d4-9ef8bb6b667e.png">

Zoomed rerun below: 
<img width="488" alt="Screenshot 2022-03-16 at 12 30 21 PM" src="https://user-images.githubusercontent.com/101671235/158639710-f8c628f3-eb6c-4ab7-b4f7-9f84d2032b06.png">

As you can see the axon area starts "spilling out" of the original axons. Because I have so many axons, for each image, I was thinking it might just be faster for me to train a new model (especially since I have 100s of images). Finally, I noticed that some axons in the branching nerves seemed to not have been counted. Is this a debugging issue, or can it be corrected by just manually counting those?

For example:
<img width="538" alt="Screenshot 2022-03-16 at 12 33 19 PM" src="https://user-images.githubusercontent.com/101671235/158640373-dcd9659e-ce8b-4b82-8ff7-4e2f8d4f842d.png">

<img width="398" alt="Screenshot 2022-03-16 at 12 33 56 PM" src="https://user-images.githubusercontent.com/101671235/158640525-cfda70b9-1680-4ea1-a064-bdda6f2eac2e.png">

I would appreciate your advice! I have attached a link to the image here. I have sent the image I am working with to your gmail.

**Reply by mathieuboudreau** [reply]
Hi @SPACEADITYA,

Quick comments, then I'll / we'll reflect more.

Personally, I think the segmentations are quite, quite good. In comparison with other external user datasets that we've handled, I think the axons that *are* segmented in your images are amongst the best we've seen overall.

1) The "bleeding out" of the axons shown in your images isn't something I've seen in other datasets. It's hard to say because I only see the segmented images, but if there is a lot of myelin touching together I could see the algorithm mistaking this for one large axon. If it only occurs a few times, it should be fairly easy to fix manually with the FSLeyes tool.

2) The missing areas (e.g. deliniated with a very straight border) are typically signs that one of the patched images (i.e. we divide your large images into small patches during the segmentation) failed. Tweaking the resolution and/or cropping the images differently could resolve this, maybe.

3) Retraining a model with your 100's of images would require these images to be *manually segmented*, which would take quite a long time practically (likely, longer than just manually fixing your images that area already segmented by ADS). On top of that, I can't see that a new model would provide significant improvement from your already very good segmentation results; the "misegmentations" that you've identified are not really failures of the model in my opinion, but the surrounding code. It would also be difficult to train a model to distinguish an area surrounded by tightly packed myelin from other axons, from one large axon wrapped with myelin, I think.

We'll take a look at your emailed datasets and report back.

**Reply by mathieuboudreau** [reply]
Also, @SPACEADITYA , 

> when I converted the image in ImageJ from 16bit to 8bit,

Thanks for pointing this out. I think we experienced this issue in the past (https://github.com/axondeepseg/axondeepseg/issues/166 and https://github.com/axondeepseg/axondeepseg/pull/243), and had resolved it, but it may have been reintroduced with recent major upgrades (https://github.com/axondeepseg/axondeepseg/pull/587).

We'll raise an issue to solve this, as you shouldn't need to do the conversion yourself, ADS should do it for you.

**Reply by Stoyan-I-A** [reply]
> 1. The "bleeding out" of the axons shown in your images isn't something I've seen in other datasets.

This might be the result of the "fill axons" button. Since the algorithm will fill closed regions of myelin which have an area lesser than 10% of the entire image area.
In the picture, we can see that the "bled out" parts are in sections where multiple axonmyelin objects touch each other, thus creating a large area closed by a connection of multiple myelin objects.

**Reply by SPACEADITYA** [reply]
Hi All, 

Now when I try processing my images, the software is taking too long to run. I haven't changed the images at all. This is what my terminal indicates. Any help would be appreciated!

<img width="1031" alt="Screenshot 2022-03-17 at 11 16 17 PM" src="https://user-images.githubusercontent.com/101671235/158930954-e70da0e5-3870-40f8-8e55-04dd14d22afb.png">
:

**Reply by mathieuboudreau** [reply]
@SPACEADITYA have you retried running it on images that previously segmented, or only on new images?

Have you maybe done anything else (re-installed AxonDeepSeg?), or maybe have a lot of softwares running (memory problem?)?

**Reply by SPACEADITYA** [reply]
Hi Mathieu, 

I am solely trying to rerun it on images that were previously segmented. Perhaps I should try a brand new image? Maybe that's the problem?

**Reply by Stoyan-I-A** [reply]
Just to diagnose the problem, could you crop your image to, let's say around 1000x1000 pixels, and try to use AxonDeepSeg to segment it?
We know that this image size shouldn't take took long to segment and should not use too much memory.

**Reply by SPACEADITYA** [reply]
Hi @Stoyan-I-A, I tried on a 1000x1000 pixel image, and it still seems to take an inordinately long time on my personal computer (Macbook Air 2020). I stopped running the program after 1 hour. Also @mathieuboudreau I'm starting to think it might be a memory problem; however, I have not observed overall slowing of other processes on my computer once I stop running fsleyes. 

I've also tried downloading AxonDeepSeg onto a much more powerful Windows computer. The installation worked perfectly. However, as I installed the ads_plugin.py on FSLeyes and clicked "yes" when asked to install permanently, FSLeyes quit. Now when I try typing "fsleyes" into the  anaconda prompt on this Windows computer, I can tell that the command is executed; however, I don't actually see the interface appear. 

I have attached a screenshot of the code in this case below: 

![image (1)](https://user-images.githubusercontent.com/101671235/159781276-c2f4f97e-e77b-4922-9a74-7e784d0cb28c.png)

I would appreciate it if someone could assist me virtually on this Windows machine as it has much more computing power than my Mac, and I'm planning on using it to segment dozens of images. 

Thank You!

**Reply by mathieuboudreau** [reply]
>I would appreciate it if someone could assist me virtually on this Windows machine as it has much more computing power than my Mac, and I'm planning on using it to segment dozens of images.

@SPACEADITYA, by any chance, have you switched between your Mac computer to your Windows computer in between the time it was working and when you started having your new problem>

**Reply by SPACEADITYA** [reply]
Hi @mathieuboudreau, I don't believe I switched between my Mac and Windows computer before encountering this new problem. I started fresh with the Windows computer and downloaded Miniconda, git, and axondeepseg--following the process layed out on AxonDeepSeg's website.

**Reply by Stoyan-I-A** [reply]
> I tried on a 1000x1000 pixel image, and it still seems to take an inordinately long time

This is not supposed to happen. It shouldn't take more than 1 minute. There might be a problem with the environment or something. What you could do is try to re-install AxonDeepSeg on a new environment (call it something like ads_venv2) and try again.

> I've also tried downloading AxonDeepSeg onto a much more powerful Windows computer.

FSLeyes is not officially supported on Windows. This might explain why it's not working properly. If you want to use your Windows machine, you have a few options:
1. Use the command line (There isn't a nice graphical interface, but it works)
2. Use the [Windows Subsystem for Linux](https://docs.microsoft.com/en-us/windows/wsl/install) so that you can run FSLeyes using the subsystem. (I don't have much experience with this but I know a member of the team uses it).
3. Use a virtual machine using VirtualBox or Vmware and the Ubuntu operating system (this will take a lot of disk space however).

**Comment by mathieuboudreau** [comment]
@SPACEADITYA would you be interested in a brief video call to get more information about your current issue? As we can't reproduce it on any of our computers, it's currently difficult to provide assistance. If so, feel free to email me, my email is on my profile.

**Reply by SPACEADITYA** [reply]
Hi @mathieuboudreau, that would be great if we could do a video call. I just emailed you. 

Thank You, 

Aditya

**Reply by SPACEADITYA** [reply]
Hi @mathieuboudreau, things are going very well since we last spoke about ADS over Zoom!

I did have one more question however. When I edit my segmentation results on ADS, I generally do not want to apply the "Fill Axons" button since that sometimes introduces more errors. As a result, I directly click the "Compute Morphometrics" button, and the morphometrics are computed based on the revised segmentation from before. The only thing is that the masks that are saved are not based on the revised segmentation. When I try resaving the axon and myelin file after I click the "Compute Morphometrics" button, the file saves as a .gz file. If the tif is packaged in the .gz file, I'm not sure how to open the .gz file since when I double click it, it decompresses into a .nii file that I do not know with which application to open with. 

Is there anyway I can simply save the myelination and axon after I segment as a tiff without clicking the "Fill Axons" button? This is the last bottleneck before my lab can start using ADS in the form of a reliable pipeline. 

Thank You So Much!

Aditya

**Reply by hermancollin** [reply]
Hi @SPACEADITYA 

If I understand correctly your problem, after having manually corrected your masks, you are not able to save the corrected versions. Have you tried using the *Save segmentations* option in the ADS menu? Also, the morphometrics feature should produce *_index.png* and *_axonmyelin_index.png* files, which are colored versions of your masks with the index of every axon overlayed. Do these look like your corrected versions or the original ones?

The *.nii.gz* files you are seeing are nifti files used internally by ADS to represent your images.

**Reply by Stoyan-I-A** [reply]
You don't have to use the `Fill Axons` button, it's just another tool that is available for you to use.

It is strange that the `compute morphometrics` button outputs .gz files. Could you check if those files weren't already present in the folder before using the `compute morphometrics` button?

**Reply by SPACEADITYA** [reply]
Hi @hermancollin, thank you for your comments. I see now that the morphometrics feature does produce index.png and axonmyelin_index.png files. Indeed, these files are the corrected version of my images. Thank you! However, the save segmentations option in the ADS menu saves my files as .gz files which I am unaware of how to open again on ADS. The functionality I especially desire is to open three things (1) my saved image (I have no problem with this with the load tiff button on ADScontrol), (2) a saved mask of previously segmented ADS myelin (axonmyelin_index instead saves both axon and myelin together), and (3) a saved mask of previously segmented ADS axon (again, axonmyelin_index instead saves both axon and myelin together), and edit each mask individually again if needed. After this additional edit, I would like to click the compute morphometrics button again and save the revised morphometrics elsewhere. Is this possible? 

And @Stoyan-I-A, thank you for your comments. I see that the compute morphometrics button does not output .gz files. As you predicted, those files were already there before using the compute morphometrics button. Thank you! I have attached images below showing (1) the outputs after clicking "Apply ADS Prediction Model" and (2) the outputs after clicking "Compute morphometrics" button. The outputs include my input tiff. 

<img width="349" alt="Screen Shot 2022-05-06 at 9 52 44 AM" src="https://user-images.githubusercontent.com/101671235/167146680-cd95ae02-cb75-49b0-a2b8-de841c79b4ed.png">

<img width="348" alt="Screen Shot 2022-05-06 at 9 56 48 AM" src="https://user-images.githubusercontent.com/101671235/167147470-d989a08b-237e-41e8-be2c-7f00b0cdbbae.png">

**Comment by hermancollin** [comment]
@SPACEADITYA I am trying to reproduce your issue. To be clear, the `.nii.gz` does not contain your segmentation. ADS will never output segmentations as nifti files: the content of this file is your original image, saved in another format. 

So, if I understand correctly the issue here, you aren't able to save your corrected axon/myelin masks? I am trying to reproduce your issue but saving corrected masks works on my side. Could you please follow this workflow and tell us where it fails:

1. Open your image (`Load PNG or TIF file` button)
2. Apply the model (`Apply ADS prediction model` button) which will save the segmentations alongside your image, or load your previously segmented masks (`Load existing mask` button).
3. Go in Edit mode (`Tools/Edit mode`) and make some changes.
![Screenshot_20220506_102929](https://user-images.githubusercontent.com/83031821/167153738-4ffbfaf8-a595-48a5-b970-597d1b8a650d.png)
4. Click on `Save segmentation` and choose a directory to save them. At this point, your corrected segmentations should be saved in png format in the directory you specified (at least it works on my side).
5. For morphometrics, when you click on `Compute morphometrics`, it prompts you for a directory to save morphometrics files (excel file + 2 index images). If you use this feature multiple times and always specify the same directory, ADS will overwrite your files, so I would suggest moving all your morphometrics files in a separate folder inbetween every computation, so that nothing is ever overwritten by accident.

Also, if you close everything and load your image again in fsleyes, using the `Load existing mask` button, you should be able to load your revised segmentations previously saved with `Save segmentation` in step 4.

**Reply by hermancollin** [reply]
Also I should mention that the `Save segmentation` button will overwrite previous files if you save them in the same directory, similarly to the morphometrics feature, so you should also be wary of not overwriting them by accident.

**Comment by SPACEADITYA** [comment]
Hello @hermancollin, 

Thank you for your response. I will describe below the challenge I am having. 

I have created a new folder in my Desktop entitled "Illustrative Example 2." In this folder, I have inserted the only tiff file that I am going to upload into the ADS program. 

![Uploading Screen Shot 2022-05-08 at 4.01.09 PM.png…]()

At this point, I adjust my zoom factor (0.1) and my micron/pixel ratio 0.1369. And now I run ADS to get the following output in that same original folder from which I had pulled my tiff file. 

<img width="1294" alt="Screen Shot 2022-05-08 at 4 02 24 PM" src="https://user-images.githubusercontent.com/101671235/167313699-adf4287d-d575-4530-9048-70c12ce3d947.png">

<img width="836" alt="Screen Shot 2022-05-08 at 4 02 49 PM" src="https://user-images.githubusercontent.com/101671235/167313721-fdbd16e6-ef14-40ad-96ad-24ece6ee6784.png">

I see that ADS has saved this original segmented image in that folder as a png file. I am now going to edit the largest axon out using edit mode. Now my GUI looks like this (you can see that the largest axon towards the top left of the image has been removed). 

<img width="1294" alt="Screen Shot 2022-05-08 at 4 04 24 PM" src="https://user-images.githubusercontent.com/101671235/167313790-821a7287-6ecd-4758-8b31-54981948bbbb.png">

I am now going to try to save the myelin mask and axon mask separately. I am hoping the file that is saved contains the modification I made to that axon. Once I click save, I get a popup that asks me to overwrite existing file. The path where this is saved looks really different from my original folder on the desktop. 

<img width="500" alt="Screen Shot 2022-05-08 at 4 05 44 PM" src="https://user-images.githubusercontent.com/101671235/167313839-dc801d99-474d-46e0-8eec-e55f6425bed2.png">

I select no, and instead select "save to new file." I then chose the original folder that contained the tiff and now the segmentations from the original time I pressed "Apply ADS prediction model." 

<img width="844" alt="Screen Shot 2022-05-08 at 4 06 54 PM" src="https://user-images.githubusercontent.com/101671235/167313870-f8d86b62-0b93-4d28-be2f-1eb78320d6dc.png">

As you can see now there are two new outputs in that original folder. Both are saved as seg-axon-copy.gz or seg-mylein-copy.gz. However, they are both .gz files. I am not sure why it is saving them as .gz files. I cannot import these files back into ADS and make further modifications should I desire to do so. 

I would appreciate your help!

<img width="837" alt="Screen Shot 2022-05-08 at 4 08 21 PM" src="https://user-images.githubusercontent.com/101671235/167313911-dfc55d56-0dcc-4f17-abb6-ce5cd760244c.png">

**Reply by Stoyan-I-A** [reply]
I think I understand what's going on. 
Are you using the "Save" icon on FSLeyes to save your images (the one right next to the image name on the `Overlay list`)? That would explain why they are saved as nii.gz files, as the default FSLeyes software uses this save format.

Please use the "Save segmentation" button on the plugin to avoid this issue. Unfortunately, this means that you can't really "save individual masks" but we do save split images of the axon and myelin masks which should do the job.

**Reply by hermancollin** [reply]
Ah yes, I think @Stoyan-I-A is right. I was able to reproduce this behavior using the "floppy disk" save button. The `Save segmentation` button in the ADScontrol pannel should have a different prompt.

**Reply by SPACEADITYA** [reply]
@Stoyan-I-A & @hermancollin, yes this works! Thank you so much!
